clients/githubrepo/branches.go

```diff
@@ -534,37 +557,49 @@ func translateRequiredStatusRepoRule(base *clients.BranchProtectionRule, rule *r
+	if base.AllowDeletions == nil || (translated.AllowDeletions != nil && !*translated.AllowDeletions) {
 		base.AllowDeletions = translated.AllowDeletions
 	}
-	if base.AllowForcePushes == nil || translated.AllowForcePushes != nil && !*translated.AllowForcePushes {
+	if base.AllowForcePushes == nil || (translated.AllowForcePushes != nil && !*translated.AllowForcePushes) {
 		base.AllowForcePushes = translated.AllowForcePushes
 	}
-	if base.EnforceAdmins == nil || translated.EnforceAdmins != nil && !*translated.EnforceAdmins {
+	if base.EnforceAdmins == nil || (translated.EnforceAdmins != nil && !*translated.EnforceAdmins) {
 		// this is an over simplification to get preliminary support for repo rules merged.
 		// A more complete approach would process all rules without bypass actors first,
 		// then process those with bypass actors. If no settings improve (due to rule layering),
 		// then we can ignore the bypass actors.
 		// https://github.com/ossf/scorecard/issues/3480
 		base.EnforceAdmins = translated.EnforceAdmins
 	}
-	if base.RequireLastPushApproval == nil || readBoolPtr(translated.RequireLastPushApproval) {
-		base.RequireLastPushApproval = translated.RequireLastPushApproval
+	if base.RequireLastPushApproval != nil || translated.RequireLastPushApproval != nil {
+		base.RequireLastPushApproval = initializedBoolRef(
+			readBoolPtr(base.RequireLastPushApproval) || readBoolPtr(translated.RequireLastPushApproval),
+		)
 	}
-	if base.RequireLinearHistory == nil || readBoolPtr(translated.RequireLinearHistory) {
-		base.RequireLinearHistory = translated.RequireLinearHistory
+	if base.RequireLinearHistory != nil || translated.RequireLinearHistory != nil {
+		base.RequireLinearHistory = initializedBoolRef(
+			readBoolPtr(base.RequireLinearHistory) || readBoolPtr(translated.RequireLinearHistory),
+		)
 	}
-	mergePullRequestReviews(&base.RequiredPullRequestReviews, &translated.RequiredPullRequestReviews)
 	mergeCheckRules(&base.CheckRules, &translated.CheckRules)
+	base.RequiredPullRequestReviews = mergePullRequestReviews(
+		base.RequiredPullRequestReviews, translated.RequiredPullRequestReviews)
 }
 
 func mergeCheckRules(base, translated *clients.StatusChecksRule) {
-	if base.UpToDateBeforeMerge == nil || readBoolPtr(translated.UpToDateBeforeMerge) {
-		base.UpToDateBeforeMerge = translated.UpToDateBeforeMerge
+	if base.UpToDateBeforeMerge != nil || translated.UpToDateBeforeMerge != nil {
+		base.UpToDateBeforeMerge = initializedBoolRef(
+			readBoolPtr(base.UpToDateBeforeMerge) || readBoolPtr(translated.UpToDateBeforeMerge),
+		)
 	}
-	if base.RequiresStatusChecks == nil || readBoolPtr(translated.RequiresStatusChecks) {
-		base.RequiresStatusChecks = translated.RequiresStatusChecks
+	if base.RequiresStatusChecks != nil || translated.RequiresStatusChecks != nil {
+		base.RequiresStatusChecks = initializedBoolRef(
+			readBoolPtr(base.RequiresStatusChecks) || readBoolPtr(translated.RequiresStatusChecks),
+		)
```

Comment:
I guess I don't see what this new version is doing that the old code didnt do.

Reply:
Ref: https://github.com/ossf/scorecard/pull/3499#discussion_r1418033642